sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -407,6 +559,411 @@ class GpuDynamicPartitionDataWriter(
+    // invoke `GpuDynamicPartitionDataSingleWriter`.write,
+    // single writer will take care of the unclosed writers and the pending caches
+    // in `concurrentWriters`
+    super.write(cb, Some(concurrentWriters))
+  }
+
+  def writeWithConcurrentWriter(cb: ColumnarBatch): Unit = {
+    this.write(cb)
+  }
+
+  /**
+   * Write an iterator of column batch.
+   *
+   * @param iterator the iterator of column batch
+   */
+  override def writeWithIterator(iterator: Iterator[ColumnarBatch]): Unit = {
+    // 1: try concurrent writer
+    while (iterator.hasNext && !fallBackToSortBased) {
+      // concurrently write and update the `concurrentWriters` map
+      // the `` will be updated
+      writeWithConcurrentWriter(iterator.next())
+    }
+
+    // 2: fall back to single writer
+    // Note single writer should restore writer status and handle the pending caches
+    if (fallBackToSortBased) {
+      // concat the put back batch and un-coming batches
+      val newIterator = Iterator.single(currentFallbackColumnarBatch) ++ iterator
+      // sort the all the batches in `iterator`
+      val sortIterator: GpuOutOfCoreSortIterator = getSorted(newIterator)
+      while (sortIterator.hasNext) {
+        // write with sort-based single writer
+        writeWithSingleWriter(sortIterator.next())
+      }
+    }
+  }
+
+  /**
+   * Sort the input iterator by out of core sort
+   *
+   * @param iterator the input iterator
+   * @return sorted iterator
+   */
+  private def getSorted(iterator: Iterator[ColumnarBatch]): GpuOutOfCoreSortIterator = {
+    val gpuSortOrder: Seq[SortOrder] = spec.sortOrder
+    val output: Seq[Attribute] = spec.exec.output
+    val sorter = new GpuSorter(gpuSortOrder, output)
+    val cpuOrd = new LazilyGeneratedOrdering(sorter.cpuOrdering)
+
+    // use noop metrics below
+    val sortTime = NoopMetric
+    val peakDevMemory = NoopMetric
+    val opTime = NoopMetric
+    val outputBatch = NoopMetric
+    val outputRows = NoopMetric
+    val spillCallback = new SpillCallback {
+      override def apply(from: StorageTier, to: StorageTier, amount: Long): Unit = {
+      }
+
+      override def semaphoreWaitTime: GpuMetric = NoopMetric
+    }
+    val targetSize = GpuSortExec.targetSize(spec.exec.conf)
+    // out of core sort the entire iterator
+    GpuOutOfCoreSortIterator(iterator, sorter, cpuOrd, targetSize,
+      opTime, sortTime, outputBatch, outputRows,
+      peakDevMemory, spillCallback)
+  }
+
+  /**
+   * concurrent write the columnar batch
+   * Note: if new partitions number in `cb` plus existing partitions number is greater than
+   * `maxWriters` limit, will put back the whole `cb` to 'single writer`
+   *
+   * @param cb the columnar batch
+   */
+  override def write(cb: ColumnarBatch): Unit = {
+    assert(isPartitioned)
+    assert(!isBucketed)
+
+    if (cb.numRows() == 0) {
+      // intend to write meta only
+      // TODO https://github.com/NVIDIA/spark-rapids/issues/6453
```

Comment:
Is this even a TODO? If the batch has no rows, then what file would we write it to? Seeing how we have nothing to get the partitions from.

Reply:
The current behavior of branch 22.10 when writing empty data:
Will never invoke `dataWriter.write` because `iterator.hasNext` is always false.
See [code link](https://github.com/NVIDIA/spark-rapids/blob/branch-22.10/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatWriter.scala#L310): 
```
val dataWriter = 
......
     new GpuDynamicPartitionDataWriter
......
        while (iterator.hasNext) {       // false
          dataWriter.write(iterator.next())   // never invoke
        }
```

My test code:
```
def test_empty_data():
    schema = StructType([
        StructField("c1", StringType()),
        StructField("c2", IntegerType()),
        StructField("c3", IntegerType()),
    ])
    data = []
    with_gpu_session(
        lambda spark: spark.createDataFrame(
            SparkContext.getOrCreate().parallelize(data, numSlices=2),
            schema).write.mode("overwrite").partitionBy('c1', 'c2').parquet("/tmp/test-empty")
    )
```

So the output file is empty and has no metadata.
```
ls /tmp/test-empty
_SUCCESS
```

The code `if (cb.numRows() == 0) {  }`  is unreachable now,  let's keep it.
After #6453 is solved, I assume we should add some code here to write metadata.